PBJS 1.0: Responsive and conditional adunits

bretg opened the pull request:
Here's the text for the Responsive and conditional adunits. @rmloveland - would like guidance on where to plug these into the site navigationally.

rmloveland commented on dev-docs/feature-conditional.md:
```diff
@@ -0,0 +1,239 @@
+---
+layout: page
+title: Conditional Ad Units - <font color="red">NEW!</font>
```
We try to name the files to match the titles - in this case would you mind renaming the file to `conditional-ad-units.md`?

rmloveland commented on dev-docs/feature-conditional.md:
```diff
@@ -0,0 +1,239 @@
+---
+layout: page
+title: Conditional Ad Units - <font color="red">NEW!</font>
+description: Using labels for conditional bids
```
This feature is referred to alternately as "conditional ad units" and "conditional bids".  I think we should standardize on one.

My vote is for "conditional ad unit" since the logic is captured inside the ad unit, and Prebid in general already calls a lot of things "bids" and makes it difficult to draw a clear distinction between client side "bids" and the actual "bids" returned by the server.

rmloveland commented on dev-docs/feature-conditional.md:
```diff
@@ -0,0 +1,239 @@
+
+<div class="bs-docs-section" markdown="1">
+
+# Conditional Bids and AdUnits
```
See above re: "conditional ad units"

rmloveland commented on dev-docs/feature-conditional.md:
```diff
@@ -0,0 +1,239 @@
+
+# Conditional Bids and AdUnits
+
+Prebid 1.0 introduced the `sizeConfig` feature primarily for [responsive ad designs]({{site.baseurl}}/dev-docs/feature-responsive.html), focusing on constraining sizes based on device.
```
Now that 1.0 is out, I think we don't need to refer to it specifically, my preference would be something like 

<blockquote>
The <code>sizeConfig</code> property is used primarily for responsive ad designs ...  It supports a number of additional features, including: ...
</blockquote>

rmloveland commented on dev-docs/feature-conditional.md:
```diff
@@ -0,0 +1,239 @@
+
+1. What if some bidders should be skipped for some devices?
+1. What if some bidders have different parameters for different devices?
+1. What if some AdUnit auctions should be skipped entirely for some devices?
```
When I see AdUnit I think the text may be referring to a type captured in the code with specific properties, e.g., in developer docs.  If we're referring to an ad unit in prose, I think we should prefer "ad unit".

rmloveland commented on dev-docs/feature-conditional.md:
```diff
@@ -0,0 +1,239 @@
+
+By supporting these scenarios, header bidding can be more efficient - the browser can send bids to a more surgical set of bidders based on device size or other attributes the page code can create.
+
+Before going into how these scenarios work, a few words about the design of Prebid's approach. With Prebid 1.0,
```
Same comment as above re: referring explicitly to Prebid 1.0 (text like this is timely now but will have to be removed later when 1.0 is no longer new, and we are likely to forget to remove it).

rmloveland replied:
I think this paragraph from "Before going into how" ... to "then used by certain AdUnits and bids" can be deleted.  What it tells me I can learn by example below, I think.

rmloveland commented on dev-docs/feature-responsive.md:
```diff
@@ -0,0 +1,80 @@
+
+## Further Reading
+
++ [Other conditional AdUnit scenarios]({{site.baseurl}}/dev-docs/feature-conditional.html)
```
I think we should add a link to the [API reference's explanation of using `labels` to show responsive ads](http://prebid.org/dev-docs/publisher-api-reference.html#setConfig-Configure-Responsive-Ads).

rmloveland commented on dev-docs/feature-responsive.md:
```diff
@@ -0,0 +1,80 @@
+---
+layout: page
+title: Responsive Ad Sizes - <font color="red">NEW!</font>
```
Let me know if I'm misunderstanding, but I think this page's content basically already exists [in the `sizeConfig`/ `labels` section of the API reference](http://prebid.org/dev-docs/publisher-api-reference.html#setConfig-Configure-Responsive-Ads).

I think instead of adding this page we should link users there.

rmloveland commented on dev-docs/feature-conditional.md:
```diff
@@ -0,0 +1,239 @@
+
+## Further Reading
+
++ [Responsive ad designs]({{site.baseurl}}/dev-docs/feature-responsive.html)
```
I think this should be a link to [the API reference's explanation of using labels to show responsive ads](http://prebid.org/dev-docs/publisher-api-reference.html#setConfig-Configure-Responsive-Ads)

rmloveland commented on dev-docs/feature-conditional.md:
```diff
@@ -0,0 +1,239 @@
+## Further Reading
+
++ [Responsive ad designs]({{site.baseurl}}/dev-docs/feature-responsive.html)
++ [Information about formulating MediaQueries](https://developer.mozilla.org/en-US/docs/Web/CSS/Media_Queries/Using_media_queries)
```
I think the link's title should be the same as the target page: "Using Media Queries".

rmloveland commented on dev-docs/feature-conditional.md:
```diff
@@ -0,0 +1,239 @@
+
+The basic steps are:
+
+1. Build up an array of labels from two sources: either as an output of sizeConfig or as an optional argument to requestBids().
```
We might want to say "from one of two sources" - later in the sentence you do make that clear!  But to avoid confusion until the reader gets there.

rmloveland commented on dev-docs/feature-conditional.md:
```diff
@@ -0,0 +1,239 @@
+The basic steps are:
+
+1. Build up an array of labels from two sources: either as an output of sizeConfig or as an optional argument to requestBids().
+1. Apply label targeting to AdUnits or specific bids.
```
Same as above re: "ad unit" in prose vs `AdUnit` type in dev docs/etc.

rmloveland commented on dev-docs/feature-conditional.md:
```diff
@@ -0,0 +1,239 @@
+1. Build up an array of labels from two sources: either as an output of sizeConfig or as an optional argument to requestBids().
+1. Apply label targeting to AdUnits or specific bids.
+
+Let's jump into an example.
```
I think this should say something like "See below for examples", followed by a Table of Contents for the page using the snippet below.  That way the reader can get an overview of the links to various use cases and jump to the one they want.  It will also make it easier to link directly to those sections in emails/Slack/etc.

```
* TOC
{:toc}
```

rmloveland commented on dev-docs/feature-conditional.md:
```diff
@@ -0,0 +1,239 @@
+Say a particular bidder is focused on mobile phone demand, so it's really not worthwhile for either party
+to send them requests from display or tablets.
+
+We'll start with how to set up the labels from sizeConfig:
```
`sizeConfig` should use code font since it's a variable name.

rmloveland commented on dev-docs/feature-conditional.md:
```diff
@@ -0,0 +1,239 @@
+
+## Use Case: What if some bidders should be skipped for some devices?
+
+Say a particular bidder is focused on mobile phone demand, so it's really not worthwhile for either party
```
I think we should remove "for either party" since the beginning of the sentence only refers to the bidder, and it's just as clear to say "so it's not really worthwhile to send them requests ..."

rmloveland commented on dev-docs/feature-conditional.md:
```diff
@@ -0,0 +1,239 @@
+
+{% endhighlight %}
+
+Here, labels are applied for each of the 3 screen sizes that can later be used in
```
IMO it would be nice to make it clearer here that:

- you have just set up labels in `setConfig`
- now you need to label your ad units to match, plus add some logic with `labelAny` and friends

rmloveland commented on dev-docs/feature-conditional.md:
```diff
@@ -0,0 +1,239 @@
+
+## Use case: What if some bidders have different parameters for different devices?
+
+For reporting and targeting purposes, Publishers and SSPs sometimes break out different 'placements' for different platforms.
```
Can you write more about the quotes around placements?  Do you mean to imply that "placement" is a new term for the reader?  If so we should define it briefly.

Or do you mean that it can mean a few different things depending on context?  If so, can we explicitly share what some of those are?

rmloveland commented on dev-docs/feature-conditional.md:
```diff
@@ -0,0 +1,239 @@
+| Display | 1111 |
+| Phones and tablets | 2222 |
+
+Assuming the same sizeConfig as in the first use case above, the AdUnit would contain bids for both
```
Same as above re: `sizeConfig` and "ad unit".

rmloveland commented on dev-docs/feature-conditional.md:
```diff
@@ -0,0 +1,239 @@
+| Phones and tablets | 2222 |
+
+Assuming the same sizeConfig as in the first use case above, the AdUnit would contain bids for both
+placements, but the conditional `labelAny` is added to them both:
```
We should add a couple of words about what `labelAny` does, e.g., "fires if one or more of the labels matches" (or something even better, IDK)

rmloveland commented on dev-docs/feature-conditional.md:
```diff
@@ -0,0 +1,239 @@
+
+{% endhighlight %}
+
+Example of how this works:
```
We could just say "how this works" as in the previous section.

rmloveland commented on dev-docs/feature-conditional.md:
```diff
@@ -0,0 +1,239 @@
+Example of how this works:
+
+1. Users with a screen width of 1000px will cause the second mediaQuery to fire.
+1. This rule puts "tablet" into the label array and a list of sizes into sizesSupported.
```
Which rule is "this rule"?

`sizesSupported` should use code font.

rmloveland commented on dev-docs/feature-conditional.md:
```diff
@@ -0,0 +1,239 @@
+
+Example of how this works:
+
+1. Users with a screen width of 1000px will cause the second mediaQuery to fire.
```
I vote we refer to it as a "media query" since it's not a proper noun.  Also that's what [the Mozilla docs do](https://developer.mozilla.org/en-US/docs/Web/CSS/Media_Queries/Using_media_queries).

rmloveland commented on dev-docs/feature-conditional.md:
```diff
@@ -0,0 +1,239 @@
+
+1. Users with a screen width of 1000px will cause the second mediaQuery to fire.
+1. This rule puts "tablet" into the label array and a list of sizes into sizesSupported.
+1. Then the AdUnit is processed:
```
Same as above re: "ad unit".

rmloveland commented on dev-docs/feature-conditional.md:
```diff
@@ -0,0 +1,239 @@
+    1. The first bid requires that the label "display" be present in the array. It's not, so that bid is skipped.
+    1. The second bid requires that either "phone" or "tablet" be present. Since tablet is in the label array, that bid is activated and the correct placement is sent to bidderA.
+
+## Use Case: What if some AdUnit auctions should be skipped entirely for some devices?
```
Same as above re: "ad unit"

rmloveland commented on dev-docs/feature-conditional.md:
```diff
@@ -0,0 +1,239 @@
+## Use Case: What if some AdUnit auctions should be skipped entirely for some devices?
+
+Say there's a responsive page where one of the AdUnits only supports larger sizes, so it doesn't make sense
+on phones. We really don't want to use that AdUnit for mobile users. Wouldn't it be great if `labelAny` and `labelAll` could be applied to the entire AdUnit? They can. Here's an example:
```
We should briefly describe `labelAny` and `labelAll`.

Same as above re: "ad unit".

Can you add some explanation around why it would be great if `labelAny` and `labelAll` could be applied to the entire ad unit?  Are you implying that the config below would only be possible using them?  If so can we explain?

rmloveland commented on dev-docs/feature-conditional.md:
```diff
@@ -0,0 +1,239 @@
+## Use Case: What if some bid requests apply only to users originating certain countries? 
+
+Labels aren't constrained to describing device size -- they can be used for any condition the page wants
+to define. To support this, labels can now be passed into the `requestBids()` function as an argument.
```
Can you please link `requestBids` (and any other API methods I may have missed) to [its entry in the API reference](http://prebid.org/dev-docs/publisher-api-reference.html#module_pbjs.requestBids)?

rmloveland replied:
Can we change "can now be passed" to "can be passed" since "now" is time-bound but we'd rather not come back and edit this later, and the recent API update will not be recent for long.

rmloveland commented on dev-docs/feature-conditional.md:
```diff
@@ -0,0 +1,239 @@
+
+## Use Case: What if some bid requests apply only to users originating certain countries? 
+
+Labels aren't constrained to describing device size -- they can be used for any condition the page wants
```
I'm a little curious about the claim "they can be used for any condition the page wants to define".  It seems overly broad.

Would it be closer to correct to say something like "almost any condition ..."?

rmloveland commented on dev-docs/feature-conditional.md:
```diff
@@ -0,0 +1,239 @@
+requestBids(reqArgs);
+{% endhighlight %}
+
+Then this label can be applied to conditions in the AdUnit just like labels that originate from sizeConfig. E.g.
```
`sizeConfig` should use code font.

rmloveland commented on dev-docs/feature-conditional.md:
```diff
@@ -0,0 +1,239 @@
+{% endhighlight %}
+
+This example shows that the 'euroMobileBidder' is only interested in receiving bids that have **both**
+labels: both "eur" as passed into requestBids and "phone" as created by sizeConfig.
```
We should remove the second use of "both".  Alternatively, this info would be easier to consume in an unordered list, e.g.,

<blockquote>
that have both labels:
- eur as ...
- phone as ...
</blockquote>

rmloveland commented on dev-docs/feature-conditional.md:
```diff
@@ -0,0 +1,239 @@
+
+## Syntax Notes
+
+Probably you've gotten the gist of the feature by now, but it's still worth laying out the syntax a little more. In Prebid 1.0, label targeting is constrained to two relatively simple operators: `labelAny` and `labelAll`.
```
I strongly believe this information should be added to [the `labels` section of the API reference](http://prebid.org/dev-docs/publisher-api-reference.html#sizeConfig-Labels) since it is reference information (not "how to" like the rest of this page).

In other words, the API reference page should contain "how stuff works", with this page being "how to use it for this use case".
Random notes on text:

First sentence makes assumptions about the reader that may not be true.

We should not refer to "Prebid 1.0" since it's timely and will need to be edited out relatively soon.

This section should be called "Label Targeting" IMO, and it could start from where you write "Label targeting is constrained to "

Also we should not describe them as "relatively simple", that's a value judgment IMO.  We can try to make the examples simple and clear though!

An important question I have is: how do you debug why your labels logic isn't matching/firing?  If you have some docs or know-how on that that you could contribute (which IMO should also go in the API reference) that would be awesome.

rmloveland commented on dev-docs/feature-conditional.md:
```diff
@@ -0,0 +1,239 @@
+// Either A or B can be defined in the label array to activate this bid or AdUnit
+{% endhighlight %}
+
+With the `labelAll` conditional, Every element of the target array must match an element of the label array in order for the condition to be true.
```
"Every" doesn't need a capital.

rmloveland commented on dev-docs/feature-conditional.md:
```diff
@@ -0,0 +1,239 @@
+Probably you've gotten the gist of the feature by now, but it's still worth laying out the syntax a little more. In Prebid 1.0, label targeting is constrained to two relatively simple operators: `labelAny` and `labelAll`.
+
+With the `labelAny` conditional, just one label has to match for the condition to be true.
+{% highlight bash %}
```
Did JS highlight look weird?

rmloveland commented on dev-docs/feature-conditional.md:
```diff
@@ -0,0 +1,239 @@
+With the `labelAny` conditional, just one label has to match for the condition to be true.
+{% highlight bash %}
+labelAny: ["A", "B"]
+// Either A or B can be defined in the label array to activate this bid or AdUnit
```
I think this comment should be part of the descriptive text above.

rmloveland replied:
See above re: "ad unit".

rmloveland commented on dev-docs/feature-conditional.md:
```diff
@@ -0,0 +1,239 @@
+With the `labelAll` conditional, Every element of the target array must match an element of the label array in order for the condition to be true.
+{% highlight bash %}
+labelAll: ["A", "B"] 
+// both A and B must be defined in the label array to activate this bid or AdUnit
```
Same as above re: this comment text should just be part of the description, and `AdUnit` vs. "ad unit"

rmloveland commented on dev-docs/feature-conditional.md:
```diff
@@ -0,0 +1,239 @@
+
+
+{: .alert.alert-warning :}
+**This feature only works in Prebid 1.0 and later.**
```
This header text doesn't match what we are describing in the paragraph below, which is more of a debugging note (which - yay, debugging info is great, as I said in an earlier comment).

I think we could have the note stand just fine on its own without the 1.0-specific header text.

rmloveland commented on dev-docs/feature-conditional.md:
```diff
@@ -0,0 +1,239 @@
+
+{: .alert.alert-warning :}
+**This feature only works in Prebid 1.0 and later.**
+Only one conditional may be specified on a given AdUnit or bid -- if both labelAny and labelAll are specified, only the first one will be utilized and an error will be logged to the console. It is allowable for an AdUnit to have one condition and a bid to have another.
```
We should refer to `labelAny` and friend(s) in `code` font.

Same as above re: "ad unit".

rmloveland commented on dev-docs/feature-conditional.md:
```diff
@@ -0,0 +1,239 @@
+**This feature only works in Prebid 1.0 and later.**
+Only one conditional may be specified on a given AdUnit or bid -- if both labelAny and labelAll are specified, only the first one will be utilized and an error will be logged to the console. It is allowable for an AdUnit to have one condition and a bid to have another.
+
+We may someday implement full boolean logic in this feature, but the common use cases don't demand the complexity or code size at this point.
```
Not a document review comment, but I really really really hope we don't ever make labels Turing complete. (We may be getting pretty close already.)

bretg commented:
Thanks for the thorough review Rich. Made most of the changes you've requested. Some areas to discuss:

1) AdUnit vs ad unit.  I find it clearer to think of AdUnits as the Prebid data structure and 'ad units' as the space in the page independent of Prebid. So I changed only references where the text was referring to the logical thing in the page as opposed to the technical data structure. But I get that consistency across the docs is worthwhile, so if you believe other docs are consistent, you can do a global replace if you'd like.

2) Not sure I like the position of the table of contents.  Thinking this needs to be manually created links to the use cases.

3) Wasn't aware of the 'Configure Responsive Ads' section of the API reference. I went ahead and removed the separate responsive page, but I do have reservations about the API ref as-is:
     - this reference is too big IMO and doesn't have a good table of contents
     - the text in the reference is outdated -- it refers to adding labels as part of setConfig, but that feature was moved to be part of requestBids()

rmloveland commented:
> Thanks for the thorough review Rich. Made most of the changes you've
> requested.

Thanks Bret! I know it was a lot of feedback. Something I forgot to say
was: thanks for submitting the PR, doc improvements and updates are
really appreciated.

> AdUnit vs ad unit. I find it clearer to think of AdUnits as the Prebid
> data structure and 'ad units' as the space in the page independent of
> Prebid. So I changed only references where the text was referring to
> the logical thing in the page as opposed to the technical data
> structure. But I get that consistency across the docs is worthwhile,
> so if you believe other docs are consistent, you can do a global
> replace if you'd like.

OK.

> Not sure I like the position of the table of contents. Thinking this
> needs to be manually created links to the use cases.

Hey that works too!

> Wasn't aware of the 'Configure Responsive Ads' section of the API
> reference. I went ahead and removed the separate responsive page, but
> I do have reservations about the API ref as-is:
>
> this reference is too big IMO

I'm not sure how to address this, since we would need to agree on what
the right size is. "It's too big" could mean "I just want the one thing
that I want right now and to ignore everything else", which I think is
achieved via the table of contents letting you jump to just the method
you want?

It could definitely be made shorter with a few editing passes.

Regarding the size and multi-sectional nature of `setConfig` in
particular, that section is large because we chose to move a lot of
functionality into it.

It's going to get even bigger once we work on [adding these undocumented
methods](https://github.com/prebid/prebid.github.io/issues/447)!

> and doesn't have a good table of contents

Right now we have a top-level TOC showing all the methods, and
subsections with their own mini-TOCs in cases like `setConfig` where a
method does many things.

I'm definitely open to suggestions/improvements!

> the text in the reference is outdated -- it refers to adding labels as
> part of setConfig, but that feature was moved to be part of
> requestBids()

Dang, didn't realize. Since you already have the text from this PR, do
you mind making a PR to update the appropriate part of the API
reference? If you're too busy I can also do it.

I've opened an issue here to track:

<https://github.com/prebid/prebid.github.io/issues/491>

rmloveland commented:
Also @bretg are there any other 1.0 API changes that snuck in under the wire besides https://github.com/prebid/prebid.github.io/issues/491 that we need to know about?

If you know of any would you mind please opening issues in this repo?

bretg commented:
@rmloveland - updated publisher-api-reference.md and prebid-1.0-API.md to address the evolution that occurred in the labels after that text was written.

Unfortunately because my branch hadn't been updated, the diffs showing here are much bigger than they actually are. I don't see a good way to diff between master and this PR, so am leaving the conflicts in place for now -- that's one way to highlight the actual changes made.

bretg commented:
PR https://github.com/prebid/prebid.github.io/pull/499/files contains all of these changes as well. Closing this one.
